### PR TITLE
Encode EdgeGroup records in the bulk-load path (BulkEdgeEncoder)

### LINKED ISSUE #389: Encode EdgeGroup records in the bulk-load path (BulkEdgeEncoder)
## Background

The bulk-load encoding path (`BulkEdgeEncoder.bulkEncodeAll` in `codec-java`) currently encodes only **EdgeState / EdgeIndex / EdgeCache / EdgeCount** from a `BulkLoadEdge`. It does **not** encode **EdgeGroup** (aggregation) records.

As a result, a table initialized via bulk-load has empty aggregation data until records accumulate incrementally through the online write path. This blocks onboarding new tables (and large migrations) that need aggregation available right after the initial load.

For contrast, the online path already builds group records in `EdgeMutationBuilder.buildGroupRecords()` (`core`) and writes them as HBase `Increment`s. `codec-java` has no EdgeGroup encoding at all — the `groups` metadata on a label is parsed (`LabelDTO.getGroups()`) but never consumed by any encoder.

## What this change does

`codec-java` will encode the **full EdgeGroup record bytes** for each edge:

- **row key**: `hash | directedSource | tableCode | EDGE_GROUP(-5) | direction.code | groupCode`
- **qualifier**: `groupValues` encoded with `Order.DESC` (including bucketed field values)
- **value**: this single edge's contribution — `1` for COUNT, the `valueField` value for SUM

Because many edges can contribute to the same `(row key, qualifier)`, summing those contributions into the final absolute cell value is done downstream by the external bulk-load pipeline (group-by + sum, then written to HFiles). This is exactly how `EdgeCount` already works: `codec-java` encodes the counter key per edge, and the pipeline aggregates the counts. The encoding is fully owned by `codec-java`; only the cross-edge aggregation is out of scope here.

## Design constraints

- **EdgeCache is the precedent.** EdgeCache encoding was added to `BulkEdgeEncoder` first, ahead of the external pipeline consuming it. EdgeGroup follows the same two-step approach: land the codec encoding here first, then update the pipeline separately.
- **Record-type byte `-5` collides.** `EncodedEdgeType.IMMUTABLE_INDEXED_EDGE_TYPE` (`-5`, dead code — "Currently not used anywhere") occupies the same physical record-type tag as the online `EDGE_GROUP` record type (`-5`). We should reuse `-5` as `EDGE_GROUP` in `codec-java`.
- **`bucket` support is missing in v2.** Group field definitions commonly use time/date `bucket`ing, but `codec-java`'s `Group.Field` has no `bucket` field (present only in the online model). Bucketed values are part of the qualifier bytes, so a date `Bucket` needs to be ported for parity.
- **Byte-layout parity is the main risk.** Bulk-encoded group records must be byte-identical to what the online `EdgeGroupRecordMapper` decodes (row key / qualifier layout above). The final `value` written by the pipeline is a raw 8-byte big-endian long.

## Task

- [ ] Structural: rename `EncodedEdgeType.IMMUTABLE_INDEXED_EDGE_TYPE(-5)` to `EDGE_GROUP_TYPE(-5)`; clean up `-5` references in `AbstractEdgeEncoder` / `DecodedEdge` (separate commit, no behavior change)
- [ ] Port the date `Bucket` into `codec-java` (`metadata/Bucket.java`) and add `bucket` to `Group.Field`
- [ ] Add group encoding methods to `EdgeEncoder` / `AbstractEdgeEncoder` / `BytesKeyValueEdgeEncoder` / `StringKeyFieldValueEdgeEncoder`, mirroring the cache-encoding path (row key + qualifier + per-edge value)
- [ ] Extend `BulkEdgeEncoder.bulkEncodeAll` to encode EdgeGroup records (COUNT + SUM) for active edges, iterating each group's own `directionType`
- [ ] Unit tests in `BulkEdgeEncoderTests` / `MultiEdgeBulkEdgeEncoderTests`
- [ ] Cross-module round-trip test in `core`'s `V2MultiEdgeBulkLoadTest`: decode the bulk-encoded group records via `EdgeGroupRecordMapper` and assert byte compatibility

## Done When

- Bulk-load encodes EdgeGroup records for COUNT and SUM groups, including `bucket`ed fields.
- `V2MultiEdgeBulkLoadTest` round-trips a bulk-encoded group key/qualifier through the `EdgeGroupRecordMapper` decoder and matches the expected `EdgeGroupRecord`.
- Existing tests (count/cache round-trips, backward compat when `groups` is absent) still pass.

## Notes

- Follows the repo's TDD / Tidy-First rules: the structural `-5` rename is committed separately from the behavioral encoding changes.
- **Out of scope (follow-up):** the external bulk-load pipeline that aggregates per-edge EdgeGroup contributions (group-by + sum for COUNT/SUM) and writes the absolute `long` values to HFiles. The pipeline currently has no qualifier slot for wide-row records, so end-to-end bulk-load of aggregation is not complete until that change lands.

### PR DESCRIPTION
**Pull request by eazyhozy**
## Summary

The bulk-load encoding path (`BulkEdgeEncoder.bulkEncodeAll` in `codec-java`) only encoded EdgeState/EdgeIndex/EdgeCache/EdgeCount from a `BulkLoadEdge`. It never encoded EdgeGroup (aggregation) records, so tables initialized via bulk-load had empty aggregation data until it accumulated incrementally through the online write path.

This PR ports the online (V3) group-encoding model into `codec-java` (V2) and wires it into `BulkEdgeEncoder`, so bulk-loaded edges emit EdgeGroup records too.

Closes #389

## Changes

- `metadata/Bucket.java` (new): ports V3's `date` bucket — epoch value → formatted string — used to bucket group fields (e.g. daily aggregation). Query-time `handleQueryValue` is intentionally not ported; bulk encoding only ever applies buckets to concrete property values.
- `metadata/Group.java`: `Group.Field` gains an optional `bucket` and an `applyBucket` helper.
- `code/EdgeEncoder.java` / `code/AbstractEdgeEncoder.java` / `code/BytesKeyValueEdgeEncoder.java` / `code/StringKeyFieldValueEdgeEncoder.java`: new `encodeGroupEdge`/`encodeAllGroupEdges` methods mirroring the existing EdgeCache encoding path. Encodes the full V3-compatible EdgeGroup record bytes for a single edge — row key (`hash|directedSrc|labelId|EDGE_GROUP|direction|groupCode`), qualifier (bucketed `groupValues`, `Order.DESC`, matching `EdgeGroupRecordMapper`'s `GROUP_RECORD_ORDER`), and this edge's single contribution as a raw 8-byte long (`1` for COUNT, the `valueField`'s value for SUM) — matching the raw `buffer.long` V3 uses for its HBase `Increment` cell, not an `OrderedBytes` header.
- `code/BulkEdgeEncoder.java`: wires EdgeGroup encoding into `bulkEncodeAll` for INDEXED and MULTI_EDGE labels, one row per `(group, direction)`, iterating each group's own `directionType` (not the label's) — matching V3 `EdgeMutationBuilder.buildGroupRecords`. Skips the `__count__` sentinel group, matching core's online mutation path (#396) — it delegates to the native EdgeCount query and never persists an EdgeGroup row.
- Tests: `BulkEdgeEncoderTests`, `MultiEdgeBulkEdgeEncoderTests`, `BucketTests` (codec-java), and `V2MultiEdgeBulkLoadTest` (core) — the bulk-encoded group key/qualifier/value round-trips through V3's `EdgeGroupRecordMapper` decoder byte-for-byte, the same safety net already used for EdgeCount/EdgeCache.

## Out of scope

Consuming these `EDGE_GROUP_TYPE` records downstream (aggregating per-edge contributions into absolute cell values, writing to HFiles) happens in the external bulk-load pipeline, not in this repo. This follows the same two-step precedent as EdgeCache: land the codec encoding first, update the pipeline separately.

## How to Test

```bash
./gradlew :codec-java:test
./gradlew :core:test --tests "*V2MultiEdgeBulkLoadTest*"
```

All tests pass, including the existing count/cache round-trips and the groups-absent backward-compatibility case.

## AI Assistance

- [x] This PR was written largely with AI assistance.
  - Tool / model: Claude Code (Sonnet 5)

### CODE REVIEW

File: codec-java/src/main/java/com/kakao/actionbase/v2/core/code/AbstractEdgeEncoder.java [hunk 1]
```diff
@@ -23,6 +28,7 @@ public abstract class AbstractEdgeEncoder<T> implements EdgeEncoder<T> {
 
   public static final byte INDEXED_EDGE_TYPE = EncodedEdgeType.INDEXED_EDGE_TYPE.getCode();
   public static final byte EDGE_CACHE_TYPE = EncodedEdgeType.EDGE_CACHE_TYPE.getCode();
+  public static final byte EDGE_GROUP_TYPE = EncodedEdgeType.EDGE_GROUP_TYPE.getCode();
```
Comment (eazyhozy) [inline codec-java/src/main/java/com/kakao/actionbase/v2/core/code/AbstractEdgeEncoder.java]: self-review: preserve ordering by code

```suggestion
  public static final byte EDGE_GROUP_TYPE = EncodedEdgeType.EDGE_GROUP_TYPE.getCode();
  public static final byte EDGE_CACHE_TYPE = EncodedEdgeType.EDGE_CACHE_TYPE.getCode();
```

File: codec-java/src/main/java/com/kakao/actionbase/v2/core/code/AbstractEdgeEncoder.java [hunk 2]
```diff
@@ -222,6 +228,72 @@ private static Object resolveFieldValue(
...
+
+    return groups.stream()
+        .filter(group -> !GROUP_COUNT_SENTINEL.equals(group.getGroup()))
+        .flatMap(
+            group ->
+                group.getDirectionType().getDirs().stream()
+                    .map(dir -> encodeGroupEdge(ts, src, tgt, props, dir, labelId, group)))
+        .collect(Collectors.toList());
+  }
+
```
Comment (eazyhozy) [inline codec-java/src/main/java/com/kakao/actionbase/v2/core/code/AbstractEdgeEncoder.java:291]: self-review: preserve ordering consistency of definitions

**Comment by em3s** [comment]
Is this ready for review? @eazyhozy 

As with the previous `EdgeCache` work, please verify that EdgeGroup works correctly in the internal bulkload path and share the test results.

https://github.com/kakao/actionbase/pull/309#issuecomment-4405277556

**Comment by eazyhozy** [comment]
@em3s To verify the bulk-load path end-to-end without relying on the internal pipeline, I'd like to add a test in `core` that simulates the pipeline's aggregation step: encode several edges via `BulkEdgeEncoder`, group-by (row key, qualifier) + sum their per-edge contributions, and assert the result matches what the online path (`EdgeMutationBuilder`) produces for the same edges.

This would sit alongside the existing `V2MultiEdgeBulkLoadTest` and `EdgeMutationBuilderTest` in `core`. Does this address the verification concern?

**Comment by em3s** [comment]
That's a great approach and worth adding.

However, to be completely confident, we still need to see it work end-to-end. Please run the internal pipeline test and share the results as well.

@eazyhozy

**Comment by eazyhozy** [comment]
@em3s Ran the end-to-end test on the internal pipeline — confirmed that EdgeGroup data is generated correctly through the bulk-load path.

**Comment by eazyhozy** [comment]
@em3s Follow-up with the full-scale end-to-end result. The earlier verification I posted was from a run that turned out to be incomplete: the bulk pipeline's encode step was writing the per-edge EdgeGroup contributions to HFiles without the group-by-and-sum aggregation, so same-bucket cells collapsed instead of accumulating. That aggregation is the pipeline's responsibility by design (as covered in V2BulkLoadAggregationTest), and after adding it on the pipeline side (group by row key + qualifier, sum contributions), the full end-to-end run passed: aggregated counts are correct across both whole-period and time-bucketed groups, verified against sampled anchor edges including events from 8 years back. Removing the duplicate rows also cut the bulk-loaded HBase row count by ~63%. No codec-java changes were needed — the encoder contract here (one contribution per edge, COUNT=1 / SUM=valueField, matching the online mutation path) held up as-is.

**Comment by em3s** [comment]
Good job! Noted that the internal tests have passed. @eazyhozy

Please merge after a final check. @zipdoki

**Comment by zipdoki** [comment]
Optimistic merge.
